Mathlib/Tactic/CC/Datatypes.lean

```diff
@@ -0,0 +1,653 @@
+  heqProofs : Bool
+  /-- If `fo == true`, then the expression associated with this entry is an application, and we are
+      using first-order approximation to encode it. That is, we ignore its partial applications. -/
+  fo : Bool
+  /-- number of elements in the equivalence class, it is meaningless if `e != root` -/
+  size : Nat
+  /-- The field `mt` is used to implement the mod-time optimization introduce by the Simplify
+      theorem prover. The basic idea is to introduce a counter gmt that records the number of
+      heuristic instantiation that have occurred in the current branch. It is incremented after each
+      round of heuristic instantiation. The field `mt` records the last time any proper descendant
+      of of thie entry was involved in a merge. -/
+  mt : Nat
+  deriving Inhabited
+
+/-- Stores equivalence class data associated with an expression `e`. -/
+abbrev Entries := RBExprMap Entry
+
+/-- Equivalence class data associated with an expression `e` used by AC congruence closure
+modules. -/
+structure ACEntry where
+  /-- Natural number associated to an expression. -/
+  idx : Nat
+  RLHSOccs : RBACAppsSet := ∅
+  RRHSOccs : RBACAppsSet := ∅
+  deriving Inhabited
+
+def ACEntry.ROccs (ent : ACEntry) : (inLHS : Bool) → RBACAppsSet
+  | true => ent.RLHSOccs
+  | false => ent.RRHSOccs
+
+structure ParentOcc where
+  expr : Expr
+  /-- If `symmTable` is true, then we should use the `symmCongruences`, otherwise `congruences`.
+      Remark: this information is redundant, it can be inferred from `expr`. We use store it for
+      performance reasons. -/
+  symmTable : Bool
+
+/-- Red-black sets of `ParentOcc`s. -/
+abbrev ParentOccSet := Std.RBSet ParentOcc (Ordering.byKey ParentOcc.expr compare)
+
+abbrev Parents := RBExprMap ParentOccSet
+
+inductive CongruencesKey
+  | fo (fn : Expr) (args : Array Expr) : CongruencesKey
+  | ho (fn : Expr) (arg : Expr) : CongruencesKey
+  deriving BEq, Hashable
+
+abbrev Congruences := Std.HashMap CongruencesKey (List Expr)
+
+structure SymmCongruencesKey where
+  (h₁ h₂ : Expr)
+  deriving BEq, Hashable
+
+abbrev SymmCongruences := Std.HashMap SymmCongruencesKey (List (Expr × Name))
+
+/-- Stores the root representatives of subsingletons. -/
+abbrev SubsingletonReprs := RBExprMap Expr
+
+/-- Stores the root representatives of `.instImplicit` arguments. -/
+abbrev InstImplicitReprs := RBExprMap (List Expr)
+
+abbrev TodoEntry := Expr × Expr × EntryExpr × Bool
+
+abbrev ACTodoEntry := ACApps × ACApps × DelayedExpr
+
+/-- Congruence closure state.
+This may be considered to be a set of expressions and an equivalence class over this set.
+The equivalence class is generated by the equational rules that are added to the `CCState` and
+congruence, that is, if `a = b` then `f(a) = f(b)` and so on. -/
+structure CCState extends CCConfig where
+  entries : Entries := ∅
+  parents : Parents := ∅
+  congruences : Congruences := ∅
+  symmCongruences : SymmCongruences := ∅
```

Comment:
```suggestion
  /-- Maps each expression to a set of expressions it might be congruent to. -/
  congruences : Congruences := ∅
  /-- Maps each expression to a set of expressions it might be congruent to,
  via the symmetrical relation. -/
  symmCongruences : SymmCongruences := ∅
```